tests/e2e/e2e_test.go

```diff
@@ -359,6 +384,57 @@ func (s *IntegrationTestSuite) TestIBCWasmHooks() {
 	)
 }
 
+// TestPacketForwarding sends a packet from chainA to chainB, and forwards it
+// back to chainA with a custom memo to execute the counter contract on chain A
+func (s *IntegrationTestSuite) TestPacketForwarding() {
+	if s.skipIBC {
+		s.T().Skip("Skipping IBC tests")
+	}
+	chainA := s.configurer.GetChainConfig(0)
+
+	nodeA, err := chainA.GetDefaultNode()
+	s.NoError(err)
+
+	// Instantiate the counter contract on chain A
+	contractAddr := s.UploadAndInstantiateCounter(chainA)
+
+	transferAmount := int64(10)
+	validatorAddr := nodeA.GetWallet(initialization.ValidatorWalletName)
+	// Specify that the counter contract should be called on chain A when the packet is received
+	contractCallMemo := []byte(fmt.Sprintf(`{"wasm":{"contract":"%s","msg": {"increment": {}} }}`, contractAddr))
+	// Generate the forward metadata
+	forwardMetadata := packetforwardingtypes.ForwardMetadata{
+		Receiver: contractAddr,
+		Port:     "transfer",
+		Channel:  "channel-0",
+		Next:     packetforwardingtypes.NewJSONObject(false, contractCallMemo, orderedmap.OrderedMap{}), // The packet sent to chainA will have this memo
+	}
+	memoData := packetforwardingtypes.PacketMetadata{Forward: &forwardMetadata}
+	forwardMemo, err := json.Marshal(memoData)
+	s.NoError(err)
+	// Send the transfer from chainA to chainB. ChainB will parse the memo and forward the packet back to chainA
+	nodeA.SendIBCTransfer(validatorAddr, validatorAddr, fmt.Sprintf("%duosmo", transferAmount), string(forwardMemo))
+
+	// check the balance of the contract
+	s.CheckBalance(nodeA, contractAddr, transferAmount)
+
+	// sender wasm addr
+	senderBech32, err := ibchookskeeper.DeriveIntermediateSender("channel-0", validatorAddr, "osmo")
+	var response map[string]interface{}
+	s.Eventually(func() bool {
+		response, err = nodeA.QueryWasmSmart(contractAddr, fmt.Sprintf(`{"get_count": {"addr": "%s"}}`, senderBech32))
+		if err != nil {
+			return false
+		}
+		count := response["count"].(float64)
+		// check if denom contains "uosmo"
+		return err == nil && count == 0
```

Comment:
We need to handle this TODO right? IMO we should ensure exact osmo amount received, in order to be certain that our default parameter has no fee taken.

Reply:
I think that comment is a leftover from previous code. The balance check is done above (line419) but it's not checking the denom, just the amount. I'll add the denom check.